python/lsst/ts/mtdome/mtdome_csc.py

```diff
@@ -374,71 +401,88 @@ async def _set_maximum_motion_values(self) -> None:
         for method, interval in methods_and_intervals:
             func = getattr(self, method)
-            self.status_tasks.append(
-                asyncio.create_task(self.one_status_loop(func, interval))
+            self.periodic_tasks.append(
+                asyncio.create_task(self.one_periodic_task(func, interval))
             )
 
+    async def one_periodic_task(self, method: typing.Callable, interval: float) -> None:
+        """Run one method forever at the specified interval.
+
+        Parameters
+        ----------
+        method: coro
+            The periodic method to run.
+        interval: `float`
+            The interval (sec) at which to run the status method.
+
+        """
+        try:
+            while True:
+                await method()
+                await asyncio.sleep(interval)
+        except Exception:
+            self.log.exception(f"one_periodic_task({method}) failed.")
+
     async def disconnect(self) -> None:
         """Disconnect from the TCP/IP controller, if connected, and stop the
         mock controller, if running.
         """
-        self.log.info("disconnect")
+        self.log.info("disconnect.")
 
-        # Stop polling for the status of the lower level components
-        # periodically.
-        await self.cancel_status_tasks()
+        # Stop all periodic tasks, including polling for the status of the
+        # lower level components.
+        await self.cancel_periodic_tasks()
 
-        writer = self.writer
-        self.reader = None
-        self.writer = None
+        if self.connected:
+            assert self.client is not None  # make mypy happy
+            await self.client.close()
+            self.client = None
         if self.simulation_mode == ValidSimulationMode.SIMULATION_WITH_MOCK_CONTROLLER:
             await self.stop_mock_ctrl()
-        if writer:
-            try:
-                writer.write_eof()
-                await asyncio.wait_for(writer.drain(), timeout=_TIMEOUT)
-            finally:
-                writer.close()
 
     async def start_mock_ctrl(self) -> None:
         """Start the mock controller.
 
         The simulation mode must be 1.
         """
-        self.log.info("start_mock_ctrl")
+        self.log.info("start_mock_ctrl.")
         try:
             assert (
                 self.simulation_mode
                 == ValidSimulationMode.SIMULATION_WITH_MOCK_CONTROLLER.value
             )
             if self.mock_port is not None:
+                host = tcpip.DEFAULT_LOCALHOST
                 port = self.mock_port
             else:
                 assert self.config is not None
+                host = self.config.host
                 port = self.config.port
-            self.mock_ctrl = MockMTDomeController(port)
+            self.mock_ctrl = MockMTDomeController(
+                host=host, port=port, log=self.log, name="MockMTDomeController"
+            )
```

Comment:
When you are starting your own mock controller here I very strongly suggest always using port 0 and the default host (which is likely tcpip.DEFAULT_LOCALHOST), because you don't want anybody other than the CSC talking to this internal mock controller.

I also strongly recommend eliminating self.mock_port entirely. If you want to run an *external* mock server (not managed by this CSC) then do the standard thing and set the config to point to it. That simplifies this code. You still need 2 simulation modes, but that is only to indicate which commands you can expect to be present. But it greatly simplifies the code to handle that mode.

Reply:
Turns out that there is exactly one use case for `self.mock_port`: when the mock controller is started by the unit test and the CSC needs to connect to that. This is useful for testing disconnections. In the end I did eliminate `self.mock_port` and am simply loading the test config file to get the port from there so it can be used by the mock controller.